ex9/lib/main.dart

```diff
@@ -0,0 +1,113 @@
+import 'package:flutter/material.dart';
+import 'package:cloud_firestore/cloud_firestore.dart';
+
+void main() => runApp(MaterialApp(home: FirestoreFirstDemo(),));
+
+FirestoreFirstDemoState pageState;
+
+class FirestoreFirstDemo extends StatefulWidget {
+  @override
+  FirestoreFirstDemoState createState() {
+    pageState = FirestoreFirstDemoState();
+    return pageState;
+  }
+}
+
+class FirestoreFirstDemoState extends State<FirestoreFirstDemo> {
+  final firestoreInstance =  Firestore .instance;
+  TextEditingController textEditingController = TextEditingController();
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(title: Text("Firestore Chat!")),
+      body: SingleChildScrollView(
+        physics: ScrollPhysics(),
+        scrollDirection: Axis.vertical,
+        child: Column(
+            mainAxisAlignment: MainAxisAlignment.spaceBetween,
+            children: <Widget>[
+              Container(
+                height: 450,
+                child: StreamBuilder<QuerySnapshot>(
+                  stream: Firestore.instance.collection("chat").snapshots(),
+                  builder: (BuildContext context,
+                      AsyncSnapshot<QuerySnapshot> snapshot) {
+                    if (snapshot.hasError) return Text("Error: ${snapshot.error}");
+                    switch (snapshot.connectionState) {
+                      case ConnectionState.waiting:
+                        return CircularProgressIndicator();
+                      default:
+                        return ListView(
+                          children: snapshot.data.documents
```

Comment:
Instead of building the entire array of message widgets, can you use ListView.builder? It's cleaner and more optimized to use a ListView.builder, pass the length, and render elements only when necessary.

https://api.flutter.dev/flutter/widgets/ListView/ListView.builder.html